File: Helpers/src/main/java/com/dg/helpers/ParcelHelper.java
package com.dg.helpers;

import android.os.Parcel;

import java.util.Date;

/**
 * Created by Daniel Cohen Gindi (danielgindi@gmail.com)
 */
public class ParcelHelper
{
    public static void writeString(Parcel parcel, String value)
    {
        parcel.writeByte(value == null ? (byte)0 : (byte)1);
        if (value != null)
        {
            parcel.writeString(value);
        }
    }

    public static void writeBoolean(Parcel parcel, Boolean value)
    {
        parcel.writeByte(value == null ? (byte)0 : (byte)1);
        if (value != null)
        {
            parcel.writeByte(value ? (byte) 0 : (byte) 1);
        }
    }

    public static void writeLong(Parcel parcel, Long value)
    {
        parcel.writeByte(value == null ? (byte)0 : (byte)1);
        if (value != null)
        {
            parcel.writeLong(value);
        }
    }

    public static void writeInt(Parcel parcel, Integer value)
    {
        parcel.writeByte(value == null ? (byte)0 : (byte)1);
        if (value != null)
        {
            parcel.writeInt(value);
        }
    }

    public static void writeDouble(Parcel parcel, Double value)
    {
        parcel.writeByte(value == null ? (byte)0 : (byte)1);
        if (value != null)
        {
            parcel.writeDouble(value);
        }
    }

    public static void writeFloat(Parcel parcel, Float value)
    {
        parcel.writeByte(value == null ? (byte)0 : (byte)1);
        if (value != null)
        {
            parcel.writeFloat(value);
        }
    }

    public static void writeDate(Parcel parcel, Date value)
    {
        parcel.writeByte(value == null ? (byte)0 : (byte)1);
        if (value != null)
        {
            parcel.writeLong(value.getTime());
        }
    }

    public static String readString(Parcel parcel)
    {
        return parcel.readByte() == 0 ? null : parcel.readString();
    }

    public static Boolean readBoolean(Parcel parcel)
    {
        return parcel.readByte() == 0 ? null : (parcel.readByte() == 1);
    }

    public static Long readLong(Parcel parcel)
    {
        return parcel.readByte() == 0 ? null : parcel.readLong();
    }

    public static Integer readInt(Parcel parcel)
    {
        return parcel.readByte() == 0 ? null : parcel.readInt();
    }

    public static Double readDouble(Parcel parcel)
    {
        return parcel.readByte() == 0 ? null : parcel.readDouble();
    }

    public static Float readFloat(Parcel parcel)
    {
        return parcel.readByte() == 0 ? null : parcel.readFloat();
    }

    public static Date readDate(Parcel parcel)
    {
        return parcel.readByte() == 0 ? null : new Date(parcel.readLong());
    }
}

File: Helpers/src/main/java/com/dg/helpers/ParcelHelper.kt
package com.dg.helpers

import android.os.Parcel

import java.util.Date

@Suppress("unused")
object ParcelHelper
{
    fun writeString(parcel: Parcel, value: String?)
    {
        parcel.writeByte(if (value == null) 0.toByte() else 1.toByte())
        if (value != null)
        {
            parcel.writeString(value)
        }
    }

    fun writeBoolean(parcel: Parcel, value: Boolean?)
    {
        parcel.writeByte(if (value == null) 0.toByte() else 1.toByte())
        if (value != null)
        {
            parcel.writeByte(if (value) 0.toByte() else 1.toByte())
        }
    }

    fun writeLong(parcel: Parcel, value: Long?)
    {
        parcel.writeByte(if (value == null) 0.toByte() else 1.toByte())
        if (value != null)
        {
            parcel.writeLong(value)
        }
    }

    fun writeInt(parcel: Parcel, value: Int?)
    {
        parcel.writeByte(if (value == null) 0.toByte() else 1.toByte())
        if (value != null)
        {
            parcel.writeInt(value)
        }
    }

    fun writeDouble(parcel: Parcel, value: Double?)
    {
        parcel.writeByte(if (value == null) 0.toByte() else 1.toByte())
        if (value != null)
        {
            parcel.writeDouble(value)
        }
    }

    fun writeFloat(parcel: Parcel, value: Float?)
    {
        parcel.writeByte(if (value == null) 0.toByte() else 1.toByte())
        if (value != null)
        {
            parcel.writeFloat(value)
        }
    }

    fun writeDate(parcel: Parcel, value: Date?)
    {
        parcel.writeByte(if (value == null) 0.toByte() else 1.toByte())
        if (value != null)
        {
            parcel.writeLong(value.time)
        }
    }

    fun readString(parcel: Parcel): String?
    {
        return if (parcel.readByte().toInt() == 0) null else parcel.readString()
    }

    fun readBoolean(parcel: Parcel): Boolean?
    {
        return if (parcel.readByte().toInt() == 0) null else parcel.readByte().toInt() == 1
    }

    fun readLong(parcel: Parcel): Long?
    {
        return if (parcel.readByte().toInt() == 0) null else parcel.readLong()
    }

    fun readInt(parcel: Parcel): Int?
    {
        return if (parcel.readByte().toInt() == 0) null else parcel.readInt()
    }

    fun readDouble(parcel: Parcel): Double?
    {
        return if (parcel.readByte().toInt() == 0) null else parcel.readDouble()
    }

    fun readFloat(parcel: Parcel): Float?
    {
        return if (parcel.readByte().toInt() == 0) null else parcel.readFloat()
    }

    fun readDate(parcel: Parcel): Date?
    {
        return if (parcel.readByte().toInt() == 0) null else Date(parcel.readLong())
    }
}
